File: src/strategies/idempotency.ts
import { type IdempotencyConfig, IdempotencyLevel, type RequestOptions } from "../core/types.js";

export class IdempotencyResolver {
  private config: IdempotencyConfig;
  private defaultLevel: IdempotencyLevel;

  constructor(
    config: Partial<IdempotencyConfig>,
    defaultLevel: IdempotencyLevel = IdempotencyLevel.SAFE,
  ) {
    this.config = {
      defaultSafeOperations: config.defaultSafeOperations ?? new Set(["GET", "HEAD", "OPTIONS"]),
      operationOverrides: config.operationOverrides ?? new Map(),
    };
    this.defaultLevel = defaultLevel;
  }

  getIdempotencyLevel(
    method: string,
    endpoint: string,
    _options: RequestOptions,
  ): IdempotencyLevel {
    const operationKey = `${method} ${endpoint}`;
    const override = this.findOverride(operationKey);
    if (override !== null) {
      return override;
    }

    if (this.config.defaultSafeOperations.has(method.toUpperCase())) {
      return IdempotencyLevel.SAFE;
    }

    return this.defaultLevel;
  }

  private findOverride(operationKey: string): IdempotencyLevel | null {
    if (this.config.operationOverrides.has(operationKey)) {
      return this.config.operationOverrides.get(operationKey)!;
    }

    for (const [pattern, level] of this.config.operationOverrides.entries()) {
      if (this.matchesPattern(pattern, operationKey)) {
        return level;
      }
    }

    return null;
  }

  private matchesPattern(pattern: string, operationKey: string): boolean {
    // Escape regex metacharacters in the literal parts of the pattern first, so a
    // pattern like "POST /charge(v2" can't produce an invalid RegExp (which would
    // throw) or be interpreted as an unintended regex. ":param" placeholders are
    // left intact by the escape (":" and word chars aren't metacharacters) and
    // then turned into path-segment wildcards.
    const escaped = pattern.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    const regexPattern = escaped.replace(/:[\w-]+/g, "[^/]+");
    let regex: RegExp;
    try {
      regex = new RegExp(`^${regexPattern}$`);
    } catch {
      // A pattern we still can't compile never matches, rather than crashing the
      // whole request on an idempotency lookup.
      return false;
    }
    return regex.test(operationKey);
  }

  shouldRetry(
    idempotencyLevel: IdempotencyLevel,
    _error: Error,
    attempt: number,
    maxRetries: number,
    hasIdempotencyKey: boolean,
  ): boolean {
    if (idempotencyLevel === IdempotencyLevel.UNSAFE) {
      return false;
    }

    if (idempotencyLevel === IdempotencyLevel.CONDITIONAL) {
      if (!hasIdempotencyKey) {
        return false;
      }
    }

    if (attempt >= maxRetries) {
      return false;
    }

    return true;
  }
}

File: clients/python/meridian/strategies/idempotency.py
"""Port of src/strategies/idempotency.ts."""

from __future__ import annotations

import re
from typing import Optional

from ..contract import IdempotencyConfig, IdempotencyLevel


class IdempotencyResolver:
    def __init__(
        self,
        config: Optional[IdempotencyConfig] = None,
        default_level: IdempotencyLevel = IdempotencyLevel.SAFE,
    ) -> None:
        self.config = config or IdempotencyConfig()
        self.default_level = default_level

    def get_idempotency_level(self, method: str, endpoint: str) -> IdempotencyLevel:
        operation_key = f"{method} {endpoint}"
        override = self._find_override(operation_key)
        if override is not None:
            return override
        if method.upper() in self.config.default_safe_operations:
            return IdempotencyLevel.SAFE
        return self.default_level

    def _find_override(self, operation_key: str) -> Optional[IdempotencyLevel]:
        overrides = self.config.operation_overrides
        if operation_key in overrides:
            return overrides[operation_key]
        for pattern, level in overrides.items():
            if self._matches_pattern(pattern, operation_key):
                return level
        return None

    @staticmethod
    def _matches_pattern(pattern: str, operation_key: str) -> bool:
        # Replace :param segments with a non-slash wildcard, mirroring the TS regex.
        regex_pattern = re.sub(r":[\w-]+", "[^/]+", pattern)
        return re.fullmatch(regex_pattern, operation_key) is not None
